### DISCUSSION TITLE
#include doesn't find matplot.h

### DISCUSSION THREAD

**Discussion by LucasMSpereira** [alandefreitas/matplotplusplus#364] [Q&A] [answered]
When using the [embed as subdirectory](https://github.com/alandefreitas/matplotplusplus#embed-as-subdirectory) procedure to get Matplot++, I'm able to use `#include <matplot/matplot.h>` in main.cpp, and make plots. However, this include doesn't work in other .cpp files. During building, I instead get a `C1083: 'matplot/matplot.h' no such file or directory` (everything unrelated to Matplot++ in the project still works fine). How do I fix this?

**Comment by alandefreitas** [comment]
```
add_subdirectory(matplotplusplus)
add_executable(my_target main.cpp)
target_link_libraries(my_target PUBLIC matplot)
```

links matplot to your target, which propagates the [INCLUDE_DIRECTORIES](https://cmake.org/cmake/help/latest/prop_tgt/INCLUDE_DIRECTORIES.html) property to your target, which now builds with `gcc -I path/to/matplot/include ...` (or anything equivalent to that in your compiler).

**Reply by LucasMSpereira** [reply]
I'm new to low level languages. I'm on windows 10, using the MSVC compiler with CMake. Those commands are already in the top level CMakeLists.txt of my project, and building my project has the behavior described in the initial post. Following your compiler CLI suggestion, from the [Microsoft Learn](https://learn.microsoft.com/en-us/cpp/build/reference/i-additional-include-directories?view=msvc-170) docs, I used `cl /I.\matplotplusplus\source\matplot \std:c++17 .\src\main.cpp`. Now main.cpp can't find matplot.h neither (pointing directly to matplot.h behaves the same). I'm pretty sure I need to add something to my CMakeLists.txt. Right now it looks like this:
```
cmake_minimum_required(VERSION 3.12.0)
project(ml_cpp VERSION 0.1.0 LANGUAGES C CXX)
set(CMAKE_CXX_STANDARD 17)
set(CMAKE_CXX_STANDARD_REQUIRED ON)
# package
add_subdirectory(matplotplusplus)
# my modules
add_subdirectory(knn)
add_subdirectory(kmeans)
add_subdirectory(lib)
# create main target
add_executable(mlCPP src/main.cpp)
# link everything
target_link_libraries(mlCPP PUBLIC knn kmeans lib matplot)
target_include_directories(mlCPP PUBLIC
"${CMAKE_HOME_DIRECTORY}/include"
"${CMAKE_HOME_DIRECTORY}/knn/include"
"${CMAKE_HOME_DIRECTORY}/kmeans/include"
)
```

**Reply by alandefreitas** [reply]
>  used cl /I.\matplotplusplus\source\matplot \std:c++17 .\src\main.cpp. Now main.cpp can't find matplot.h neither (pointing directly to matplot.h behaves the same)

You're going to have problems with this command anyway because you would also have to link to the libraries (which is not simple and exactly what CMake helps us with). But if the particular problem you're having with this command is the header can't be found (and not any other compiler error), then the `/I` path must be just wrong. I can't really think of anything else.

> target_link_libraries(mlCPP PUBLIC knn kmeans lib matplot)

This command seems OK. matplot should already include its directory:

https://github.com/alandefreitas/matplotplusplus/blob/156001b56efb9b79c888426a7754c3255f15b665/source/matplot/CMakeLists.txt#L96-L98

If CMake gives you no errors, you can only debug the cmake script to check what are the properties of these targets.

**Comment by LucasMSpereira** [comment, marked as answer]
I made a [post](https://discourse.cmake.org/t/header-file-of-package-not-found/8680/1) on CMake's discourse. A user there solved the problem. I needed to add `target_link_libraries(knn PUBLIC Matplot++::matplot)` in the CMakeLists.txt file of the knn module.

**Reply by alandefreitas** [reply]
Oh... I had no idea `knn` depended on `matplot` from the context. I thought it was an external library. Yes. All libraries have to link their dependencies.
